LTP: Fix test case mkdir02, mkdir04 issue

shaikshavali1 opened the pull request:
Test cases(mkdir02, mkdir04) is failed/skipped
because it exit improperly in middle of test. This is due
to single process environment. Test case is modified
to work in single process environment.

Commit:
LTP: Fix test case mkdir02 issue

Test case(testcases-kernel-mkdir-mkdir02) is failed/skipped
because it exit improperly in middle of test. This is due
to single process environment. Test case is modified
to work in single process environment.

hukoyu commented on tests/ltp/patches/fix_mkdir_mkdir02.patch:
```diff
@@ -0,0 +1,74 @@
++	int usr_secbits;
++
++	/*
++	 * if program started as a root user, avoid clearing of
```
Stars looks ugly in the comment. Please fix. Better you can use multiple line // comments

shaikshavali1 replied:
Ok, I will use the single line comments. modified the code and update patch

shaikshavali1 replied:
Done

hukoyu commented on tests/ltp/patches/fix_mkdir_mkdir02.patch:
```diff
@@ -0,0 +1,74 @@
++	/*
++	 * if program started as a root user, avoid clearing of
++	 * effective capabilities when changing from user root
++         * non-root user.
```
I think you ment to non-root user

shaikshavali1 replied:
I will correct the comment

shaikshavali1 replied:
Done

hukoyu commented on tests/ltp/patches/fix_mkdir_mkdir02.patch:
```diff
@@ -0,0 +1,74 @@
+	process environment by Removing exit code from child and 
+	wait code from parent. FORK_OR_VFORK return always 0 in
+	sgxlkl environment. Application in sgx-lkl environment
+	start with root user if this is the case Effective
```
This change is necessary to run this test case in sgx-lkl
	environment. Test case is modified to work in single
	process environment by removing exit code from child and 
	wait code from parent. FORK_OR_VFORK always returns 0 in
	sgxlkl environment. Application in sgx-lkl environment
	starts with root user and some permissions lost when user
	changed from root to non-root. Hence, test case is updated
	to handle this scenario as well.

shaikshavali1 replied:
I will update the patch description as per your comment.

shaikshavali1 replied:
Done

hukoyu reviewed (changes requested):
Mostly feedback is about comments. Grammatical and style issues.

Commit:
LTP: Fix test case mkdir04 issue

Test case(testcases-kernel-mkdir-mkdir04) is failed/skipped
because it exit improperly in middle of test. This is due
to single process environment. Test case is modified
to work in single process environment.

shaikshavali1 commented:
@hukoyu 
I pushed new commit for mkdir04 test case as well. Could you please check the new commit "https://github.com/lsds/sgx-lkl/pull/249/commits/841e9dbf68bb83ce6f0667110cf17045212a2123".

shaikshavali1 committed:
fix review comments from reviewer

shaikshavali1 commented:
@hukoyu , I addressed the review comments could you please check once.

shaikshavali1 committed:
update patch description

hukoyu reviewed (approved):
LGTM. cc @letmaik

letmaik commented on tests/ltp/patches/fix_mkdir_mkdir02.patch:
```diff
@@ -0,0 +1,61 @@
+ 	pid_t pid;
++	int usr_secbits;
++
++	// set the flag "SECBIT_NO_SETUID_FIXUP" to retain the
```
It's not clear why this patch is needed to make mkdir work. Could you add more details?
Also, please use `if(getuid() == 0)` since your intention is not comparing a boolean.

shaikshavali1 replied:
Modified the expression in if condition. Details about the patch is added in patch description.

letmaik commented on tests/ltp/patches/fix_mkdir_mkdir02.patch:
```diff
@@ -0,0 +1,61 @@
+ 	pw = SAFE_GETPWNAM("bin");
+ 	bin_uid = pw->pw_uid;
+ 	bin_gid = pw->pw_gid;
+-
```
Remove formatting change in patch.

shaikshavali1 replied:
Done

letmaik commented on tests/ltp/patches/fix_mkdir_mkdir02.patch:
```diff
@@ -0,0 +1,61 @@
+ 		if (!fail)
+ 			tst_res(TPASS, "New dir inherited GID and S_ISGID");
+ 
+-		exit(0);
```
Why was this removed?

shaikshavali1 replied:
FORK_OR_VFORK in LTP returns always 0 in sgxlkl environment because sgx-lkl support single process. Original test case create a child  and exit from child. But in sgxlkl this will cause exit from the test program before test completion.  Updated the details in patch description.

letmaik commented on tests/ltp/patches/fix_mkdir_mkdir02.patch:
```diff
@@ -0,0 +1,61 @@
+ 		SAFE_MKDIR(TESTDIR1, 0777);
+ 		SAFE_STAT(TESTDIR1, &buf);
+ 		SAFE_CHMOD(TESTDIR1, buf.st_mode | S_ISGID);
+-		exit(0);
```
Same as above.

shaikshavali1 replied:
Updated the details in patch description.

letmaik commented on tests/ltp/patches/fix_mkdir_mkdir02.patch:
```diff
@@ -0,0 +1,61 @@
+diff --git a/testcases/kernel/syscalls/mkdir/mkdir02.c b/testcases/kernel/syscalls/mkdir/mkdir02.c
```
Summary description of patch is missing. See other patches.

shaikshavali1 replied:
Done

shaikshavali1 committed:
Fixed review comments from reviewer

shaikshavali1 commented:
@letmaik, 
Hi Maik,
I addressed your review comments. Could you please check and approve.

letmaik commented on tests/ltp/patches/fix_mkdir_mkdir02.patch:
```diff
@@ -0,0 +1,61 @@
+This change is necessary to run this test case in sgx-lkl
+environment. Test case is modified to work in single
+process environment by removing exit code from child and
```
You're not removing an exit code, you're removing a call to `exit()`. What is a "wait code"? I couldn't see that in the patch.

shaikshavali1 replied:
@letmaik, The 'tst_reap_children()' function makes the process wait for all of its
children.

letmaik replied:
@hukoyu Could you please help rephrase the description to make it more clear? Thanks!

hukoyu replied:
In the original test a child process is used, and parent process waits for child process to finish, and child process calls exit to return to parent process. The test case is modified to work in sgx-lkl single
process environment by removing codes related to exit from child and
wait on parent. FORK_OR_VFORK always returns 0 in sgxlkl environment. 
In addition, application in sgx-lkl environment starts with root user and 
admin permissions lost when user changed from root to non-root user. 
Test case is updated to handle this scenario as well.

hukoyu replied:
@shaikshavali1 I reprhased the description. Please fix the formatting and length of lines.

shaikshavali1 replied:
@hukoyu,  modified the description as suggested.

shaikshavali1 replied:
@letmaik ,
Hi Mike,
Could you please re-run the failed jobs in Azure pipe line?

In Azure pipeline build, below mentioned two test cases were failed before starting the test application in enclave. These failure are not related to my commit. 

languages-java-read_file-(debug)-(run-hw)-(8-ethreads) 
basic-global_vars_test-(debug)-(run-sw)-(8-ethreads)

shaikshavali1 committed:
Fixed review comments from reviewer

shaikshavali1 committed:
update patch description

hukoyu reviewed (approved):
LGTM cc @letmaik
